File: strategies/6-branch-and-bound/TravelingSalesman.c
/**
 * Traveling Salesman Problem (TSP) - Branch & Bound Implementation in C
 * 
 * This implementation demonstrates the Branch & Bound strategy for solving TSP,
 * showcasing systematic exploration with intelligent pruning.
 * 
 * Key Features:
 * - Priority queue-based best-first search
 * - MST-based lower bound calculation
 * - Comprehensive pruning strategies
 * - Step-by-step execution tracking
 * - Performance comparison with brute force
 * 
 * Compilation: gcc -o tsp TravelingSalesman.c -lm
 * Usage: ./tsp
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <time.h>
#include <math.h>

#define MAX_CITIES 20
#define INF INT_MAX

typedef struct {
    int path[MAX_CITIES];     // Current path
    int visited[MAX_CITIES];  // Visited cities
    int cost;                 // Current path cost
    int bound;                // Lower bound estimate
    int level;                // Current level in search tree
} TSPNode;

typedef struct {
    TSPNode nodes[10000];     // Priority queue
    int size;                 // Current size
} PriorityQueue;

typedef struct {
    int path[MAX_CITIES + 1]; // Best path found
    int cost;                 // Best cost
    int nodes_explored;       // Nodes explored
    int nodes_pruned;         // Nodes pruned
    double execution_time;    // Time in milliseconds
} TSPResult;

// Global variables
int graph[MAX_CITIES][MAX_CITIES];
int n;
int best_cost;
int best_path[MAX_CITIES + 1];
int nodes_explored;
int nodes_pruned;
int verbose;

// Function prototypes
void solve_tsp();
int calculate_bound(TSPNode* node);
void pq_init(PriorityQueue* pq);
void pq_push(PriorityQueue* pq, TSPNode node);
TSPNode pq_pop(PriorityQueue* pq);
int pq_empty(PriorityQueue* pq);
TSPResult brute_force_tsp();
void print_graph();
void print_path(int* path, int length);
void demonstrate_scaling();
int next_permutation(int* arr, int n);

/**
 * Initialize priority queue
 */
void pq_init(PriorityQueue* pq) {
    pq->size = 0;
}

/**
 * Push node to priority queue (min-heap based on bound)
 */
void pq_push(PriorityQueue* pq, TSPNode node) {
    int i = pq->size++;
    pq->nodes[i] = node;
    
    // Bubble up
    while (i > 0) {
        int parent = (i - 1) / 2;
        if (pq->nodes[i].bound >= pq->nodes[parent].bound) break;
        
        TSPNode temp = pq->nodes[i];
        pq->nodes[i] = pq->nodes[parent];
        pq->nodes[parent] = temp;
        i = parent;
    }
}

/**
 * Pop minimum bound node from priority queue
 */
TSPNode pq_pop(PriorityQueue* pq) {
    TSPNode min = pq->nodes[0];
    pq->nodes[0] = pq->nodes[--pq->size];
    
    // Bubble down
    int i = 0;
    while (2 * i + 1 < pq->size) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        int smallest = left;
        
        if (right < pq->size && pq->nodes[right].bound < pq->nodes[left].bound) {
            smallest = right;
        }
        
        if (pq->nodes[i].bound <= pq->nodes[smallest].bound) break;
        
        TSPNode temp = pq->nodes[i];
        pq->nodes[i] = pq->nodes[smallest];
        pq->nodes[smallest] = temp;
        i = smallest;
    }
    
    return min;
}

/**
 * Check if priority queue is empty
 */
int pq_empty(PriorityQueue* pq) {
    return pq->size == 0;
}

/**
 * Calculate lower bound using reduced matrix approach
 */
int calculate_bound(TSPNode* node) {
    int bound = node->cost;
    
    // For each city, add minimum outgoing edge cost
    for (int i = 0; i < n; i++) {
        if (!node->visited[i] || (node->level > 0 && i == node->path[node->level - 1])) {
            int min1 = INF, min2 = INF;
            
            // Find two smallest edges from city i
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    if (graph[i][j] < min1) {
                        min2 = min1;
                        min1 = graph[i][j];
                    } else if (graph[i][j] < min2 && graph[i][j] != min1) {
                        min2 = graph[i][j];
                    }
                }
            }
            
            // Add contribution to bound
            if (node->visited[i] && node->level > 0 && i != node->path[node->level - 1]) {
                bound += min1;
            } else if (min2 != INF) {
                bound += (min1 + min2) / 2;
            } else {
                bound += min1;
            }
        }
    }
    
    return bound;
}

/**
 * Solve TSP using Branch & Bound
 */
void solve_tsp() {
    clock_t start = clock();
    
    if (verbose) {
        printf("=== Starting TSP Branch & Bound Solution ===\n");
        printf("Cities: %d, Graph size: %dx%d\n", n, n, n);
    }
    
    // Initialize
    best_cost = INF;
    nodes_explored = 0;
    nodes_pruned = 0;
    
    // Create root node
    TSPNode root;
    memset(root.visited, 0, sizeof(root.visited));
    root.path[0] = 0;
    root.visited[0] = 1;
    root.cost = 0;
    root.level = 1;
    root.bound = calculate_bound(&root);
    
    if (verbose) {
        printf("Root node bound: %d\n", root.bound);
    }
    
    // Initialize priority queue
    PriorityQueue pq;
    pq_init(&pq);
    pq_push(&pq, root);
    
    while (!pq_empty(&pq)) {
        TSPNode current = pq_pop(&pq);
        nodes_explored++;
        
        if (verbose && nodes_explored <= 20) {
            printf("Exploring node at level %d, bound: %d\n", current.level, current.bound);
            printf("  Path: ");
            for (int i = 0; i < current.level; i++) {
                printf("%d ", current.path[i]);
            }
            printf("\n");
        }
        
        // Pruning check
        if (current.bound >= best_cost) {
            nodes_pruned++;
            if (verbose && nodes_explored <= 20) {
                printf("  Pruned: bound %d >= best %d\n", current.bound, best_cost);
            }
            continue;
        }
        
        // Check if we have a complete tour
        if (current.level == n) {
            int return_cost = graph[current.path[n-1]][0];
            int total_cost = current.cost + return_cost;
            
            if (total_cost < best_cost) {
                best_cost = total_cost;
                for (int i = 0; i < n; i++) {
                    best_path[i] = current.path[i];
                }
                best_path[n] = 0; // Complete the cycle
                
                if (verbose) {
                    printf("*** New best solution found! ***\n");
                    printf("Path: ");
                    print_path(best_path, n + 1);
                    printf("Cost: %d\n", best_cost);
                }
            }
            continue;
        }
        
        // Branch to unvisited cities
        for (int city = 1; city < n; city++) {
            if (!current.visited[city]) {
                TSPNode child = current;
                child.path[child.level] = city;
                child.visited[city] = 1;
                child.level++;
                child.cost += graph[current.path[current.level - 1]][city];
                child.bound = calculate_bound(&child);
                
                // Only add to queue if bound is promising
                if (child.bound < best_cost) {
                    pq_push(&pq, child);
                } else {
                    nodes_pruned++;
                }
            }
        }
    }
    
    clock_t end = clock();
    double execution_time = ((double)(end - start)) / CLOCKS_PER_SEC * 1000.0;
    
    if (verbose) {
        printf("=== Final Results ===\n");
        printf("Best path: ");
        print_path(best_path, n + 1);
        printf("Best cost: %d\n", best_cost);
        printf("Nodes explored: %d\n", nodes_explored);
        printf("Nodes pruned: %d\n", nodes_pruned);
        printf("Execution time: %.2f ms\n", execution_time);
    }
}

/**
 * Brute force TSP solution for comparison
 */
TSPResult brute_force_tsp() {
    clock_t start = clock();
    
    int cities[MAX_CITIES];
    for (int i = 1; i < n; i++) {
        cities[i-1] = i;
    }
    
    int min_cost = INF;
    int min_path[MAX_CITIES + 1];
    int permutations = 0;
    
    do {
        permutations++;
        int cost = 0;
        int prev = 0;
        
        for (int i = 0; i < n - 1; i++) {
            cost += graph[prev][cities[i]];
            prev = cities[i];
        }
        cost += graph[prev][0]; // Return to start
        
        if (cost < min_cost) {
            min_cost = cost;
            min_path[0] = 0;
            for (int i = 0; i < n - 1; i++) {
                min_path[i + 1] = cities[i];
            }
            min_path[n] = 0;
        }
    } while (next_permutation(cities, n - 1));
    
    clock_t end = clock();
    double execution_time = ((double)(end - start)) / CLOCKS_PER_SEC * 1000.0;
    
    TSPResult result;
    for (int i = 0; i <= n; i++) {
        result.path[i] = min_path[i];
    }
    result.cost = min_cost;
    result.nodes_explored = permutations;
    result.nodes_pruned = 0;
    result.execution_time = execution_time;
    
    return result;
}

/**
 * Generate next permutation (lexicographic order)
 */
int next_permutation(int* arr, int n) {
    int i = n - 2;
    
    while (i >= 0 && arr[i] >= arr[i + 1]) {
        i--;
    }
    
    if (i < 0) return 0;
    
    int j = n - 1;
    while (arr[j] <= arr[i]) {
        j--;
    }
    
    // Swap
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
    
    // Reverse suffix
    int left = i + 1, right = n - 1;
    while (left < right) {
        temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
        left++;
        right--;
    }
    
    return 1;
}

/**
 * Print adjacency matrix
 */
void print_graph() {
    printf("    ");
    for (int i = 0; i < n; i++) {
        printf("%4d", i);
    }
    printf("\n");
    
    for (int i = 0; i < n; i++) {
        printf("%2d: ", i);
        for (int j = 0; j < n; j++) {
            if (graph[i][j] == 0) {
                printf("   -");
            } else {
                printf("%4d", graph[i][j]);
            }
        }
        printf("\n");
    }
}

/**
 * Print path array
 */
void print_path(int* path, int length) {
    for (int i = 0; i < length; i++) {
        printf("%d", path[i]);
        if (i < length - 1) printf(" -> ");
    }
    printf("\n");
}

/**
 * Demonstrate scaling behavior
 */
void demonstrate_scaling() {
    printf("\n=== Scaling Analysis ===\n");
    
    srand(42);
    int sizes[] = {4, 5, 6, 7};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
    
    printf("Problem Size | Nodes Explored | Nodes Pruned | Time (ms) | Pruning %%\n");
    printf("----------------------------------------------------------------------\n");
    
    for (int s = 0; s < num_sizes; s++) {
        n = sizes[s];
        
        // Generate random symmetric graph
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) {
                    graph[i][j] = 0;
                } else if (i < j) {
                    graph[i][j] = graph[j][i] = 10 + rand() % 40;
                } else {
                    graph[i][j] = graph[j][i];
                }
            }
        }
        
        verbose = 0;
        solve_tsp();
        
        double pruning_percent = (double)nodes_pruned / (nodes_explored + nodes_pruned) * 100.0;
        
        printf("%11d | %13d | %11d | %8.2f | %8.1f%%\n",
               n, nodes_explored, nodes_pruned, 
               0.0, pruning_percent); // Time calculation simplified for this demo
    }
    
    printf("\nKey Observations:\n");
    printf("- Pruning effectiveness increases with problem size\n");
    printf("- Branch & bound dramatically reduces search space\n");
    printf("- Good bounds are crucial for performance\n");
}

int main() {
    printf("=== Traveling Salesman Problem - Branch & Bound ===\n\n");
    
    // Test case 1: Small symmetric graph
    n = 4;
    int graph1[4][4] = {
        {0, 10, 15, 20},
        {10, 0, 35, 25},
        {15, 35, 0, 30},
        {20, 25, 30, 0}
    };
    
    // Copy to global graph
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            graph[i][j] = graph1[i][j];
        }
    }
    
    printf("Test Case 1: 4-city symmetric TSP\n");
    printf("Graph:\n");
    print_graph();
    
    verbose = 1;
    solve_tsp();
    
    // Compare with brute force
    printf("\nComparing with Brute Force:\n");
    TSPResult brute_result = brute_force_tsp();
    printf("=== Brute Force Results ===\n");
    printf("Permutations checked: %d\n", brute_result.nodes_explored);
    printf("Best path: ");
    print_path(brute_result.path, n + 1);
    printf("Best cost: %d\n", brute_result.cost);
    printf("Execution time: %.2f ms\n", brute_result.execution_time);
    
    printf("\nEfficiency Comparison:\n");
    printf("Branch & Bound: %d nodes explored, %d pruned\n", nodes_explored, nodes_pruned);
    printf("Brute Force: %d permutations checked\n", brute_result.nodes_explored);
    
    // Test case 2: Larger asymmetric graph
    printf("\n============================================================\n");
    printf("Test Case 2: 5-city asymmetric TSP\n");
    
    n = 5;
    int graph2[5][5] = {
        {0, 12, 29, 22, 13},
        {12, 0, 19, 28, 25},
        {29, 19, 0, 21, 10},
        {22, 28, 21, 0, 24},
        {13, 25, 10, 24, 0}
    };
    
    // Copy to global graph
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            graph[i][j] = graph2[i][j];
        }
    }
    
    printf("Graph:\n");
    print_graph();
    
    verbose = 0;
    solve_tsp();
    
    printf("\nBranch & Bound Results:\n");
    printf("Best path: ");
    print_path(best_path, n + 1);
    printf("Best cost: %d\n", best_cost);
    printf("Nodes explored: %d\n", nodes_explored);
    printf("Nodes pruned: %d\n", nodes_pruned);
    
    // Performance analysis
    printf("\n=== Performance Analysis ===\n");
    printf("Branch & Bound Advantages:\n");
    printf("- Optimal solution guaranteed\n");
    printf("- Systematic pruning reduces search space\n");
    printf("- Best-first search finds good solutions early\n");
    printf("- Memory efficient compared to dynamic programming\n");
    
    printf("\nWhen to use Branch & Bound for TSP:\n");
    printf("- Small to medium problem sizes (< 20 cities)\n");
    printf("- When optimal solution is required\n");
    printf("- When good bounds can be computed efficiently\n");
    printf("- Problems with strong constraint structure\n");
    
    demonstrate_scaling();
    
    return 0;
}

File: strategies/6-branch-and-bound/TravelingSalesman.java
/**
 * Traveling Salesman Problem (TSP) - Branch & Bound Implementation
 * 
 * This implementation demonstrates the Branch & Bound strategy for solving TSP,
 * an NP-hard optimization problem. The algorithm systematically explores
 * the solution space while using bounds to prune suboptimal branches.
 * 
 * Key Concepts:
 * - Branch & Bound: Systematic enumeration with pruning
 * - Lower Bound Calculation: Minimum Spanning Tree based bounds
 * - State Space Tree: Partial tour representations
 * - Pruning: Eliminating suboptimal branches early
 * 
 * Time Complexity: O(n!) worst case, but often much better with pruning
 * Space Complexity: O(n) for recursion stack
 * 
 * Real-world Applications:
 * - Route optimization for delivery services
 * - Circuit board drilling optimization
 * - DNA sequencing
 * - VLSI design
 */

import java.util.*;

public class TravelingSalesman {
    
    static class TSPNode {
        List<Integer> path;     // Current partial path
        boolean[] visited;      // Visited cities
        int cost;              // Cost of current path
        int bound;             // Lower bound estimate
        int level;             // Depth in search tree
        
        TSPNode(int n) {
            path = new ArrayList<>();
            visited = new boolean[n];
            cost = 0;
            bound = 0;
            level = 0;
        }
        
        TSPNode(TSPNode other) {
            path = new ArrayList<>(other.path);
            visited = other.visited.clone();
            cost = other.cost;
            bound = other.bound;
            level = other.level;
        }
    }
    
    private int[][] graph;
    private int n;
    private int bestCost;
    private List<Integer> bestPath;
    private int nodesExplored;
    private int nodesPruned;
    
    // Step-by-step demonstration variables
    private List<String> steps;
    private boolean verbose;
    
    public TravelingSalesman(int[][] adjacencyMatrix, boolean verbose) {
        this.graph = adjacencyMatrix;
        this.n = graph.length;
        this.bestCost = Integer.MAX_VALUE;
        this.bestPath = new ArrayList<>();
        this.nodesExplored = 0;
        this.nodesPruned = 0;
        this.steps = new ArrayList<>();
        this.verbose = verbose;
    }
    
    /**
     * Solve TSP using Branch & Bound
     */
    public TSPResult solveTSP() {
        long startTime = System.nanoTime();
        
        if (verbose) steps.add("=== Starting TSP Branch & Bound Solution ===");
        if (verbose) steps.add("Cities: " + n + ", Graph size: " + n + "x" + n);
        
        // Initialize root node
        TSPNode root = new TSPNode(n);
        root.path.add(0);  // Start from city 0
        root.visited[0] = true;
        root.level = 1;
        root.bound = calculateBound(root);
        
        if (verbose) steps.add("Root node bound: " + root.bound);
        
        // Use priority queue for best-first search
        PriorityQueue<TSPNode> pq = new PriorityQueue<>((a, b) -> a.bound - b.bound);
        pq.offer(root);
        
        while (!pq.isEmpty()) {
            TSPNode current = pq.poll();
            nodesExplored++;
            
            if (verbose && nodesExplored <= 20) {
                steps.add("Exploring node at level " + current.level + 
                         ", path: " + current.path + ", bound: " + current.bound);
            }
            
            // Pruning check
            if (current.bound >= bestCost) {
                nodesPruned++;
                if (verbose && nodesExplored <= 20) {
                    steps.add("  Pruned: bound " + current.bound + " >= best " + bestCost);
                }
                continue;
            }
            
            // Check if we have a complete tour
            if (current.level == n) {
                // Add cost to return to start
                int returnCost = graph[current.path.get(n-1)][0];
                int totalCost = current.cost + returnCost;
                
                if (totalCost < bestCost) {
                    bestCost = totalCost;
                    bestPath = new ArrayList<>(current.path);
                    bestPath.add(0); // Complete the cycle
                    
                    if (verbose) {
                        steps.add("*** New best solution found! ***");
                        steps.add("Path: " + bestPath + ", Cost: " + bestCost);
                    }
                }
                continue;
            }
            
            // Branch to unvisited cities
            for (int city = 1; city < n; city++) {
                if (!current.visited[city]) {
                    TSPNode child = new TSPNode(current);
                    child.path.add(city);
                    child.visited[city] = true;
                    child.level = current.level + 1;
                    child.cost = current.cost + graph[current.path.get(current.level-1)][city];
                    child.bound = calculateBound(child);
                    
                    // Only add to queue if bound is promising
                    if (child.bound < bestCost) {
                        pq.offer(child);
                    } else {
                        nodesPruned++;
                    }
                }
            }
        }
        
        long endTime = System.nanoTime();
        double executionTime = (endTime - startTime) / 1_000_000.0;
        
        if (verbose) {
            steps.add("=== Final Results ===");
            steps.add("Best path: " + bestPath);
            steps.add("Best cost: " + bestCost);
            steps.add("Nodes explored: " + nodesExplored);
            steps.add("Nodes pruned: " + nodesPruned);
            steps.add("Execution time: " + String.format("%.2f", executionTime) + " ms");
        }
        
        return new TSPResult(new ArrayList<>(bestPath), bestCost, 
                           nodesExplored, nodesPruned, executionTime, new ArrayList<>(steps));
    }
    
    /**
     * Calculate lower bound using minimum spanning tree approach
     */
    private int calculateBound(TSPNode node) {
        int bound = node.cost;
        
        // For unvisited cities, add minimum outgoing edge costs
        for (int i = 0; i < n; i++) {
            if (!node.visited[i] || i == node.path.get(node.path.size()-1)) {
                int min1 = Integer.MAX_VALUE, min2 = Integer.MAX_VALUE;
                
                for (int j = 0; j < n; j++) {
                    if (i != j && graph[i][j] < min1) {
                        min2 = min1;
                        min1 = graph[i][j];
                    } else if (i != j && graph[i][j] < min2 && graph[i][j] != min1) {
                        min2 = graph[i][j];
                    }
                }
                
                // For visited cities (except current), use one minimum edge
                // For unvisited cities, use average of two minimum edges
                if (node.visited[i] && i != node.path.get(node.path.size()-1)) {
                    bound += min1;
                } else {
                    bound += (min1 + min2) / 2;
                }
            }
        }
        
        return bound;
    }
    
    /**
     * Brute force solution for comparison
     */
    public TSPResult bruteForceTSP() {
        long startTime = System.nanoTime();
        
        List<Integer> cities = new ArrayList<>();
        for (int i = 1; i < n; i++) {
            cities.add(i);
        }
        
        int minCost = Integer.MAX_VALUE;
        List<Integer> minPath = new ArrayList<>();
        int permutations = 0;
        
        do {
            permutations++;
            int cost = 0;
            int prev = 0;
            
            for (int city : cities) {
                cost += graph[prev][city];
                prev = city;
            }
            cost += graph[prev][0]; // Return to start
            
            if (cost < minCost) {
                minCost = cost;
                minPath = new ArrayList<>();
                minPath.add(0);
                minPath.addAll(cities);
                minPath.add(0);
            }
        } while (nextPermutation(cities));
        
        long endTime = System.nanoTime();
        double executionTime = (endTime - startTime) / 1_000_000.0;
        
        List<String> bruteSteps = Arrays.asList(
            "=== Brute Force Results ===",
            "Permutations checked: " + permutations,
            "Best path: " + minPath,
            "Best cost: " + minCost,
            "Execution time: " + String.format("%.2f", executionTime) + " ms"
        );
        
        return new TSPResult(minPath, minCost, permutations, 0, executionTime, bruteSteps);
    }
    
    private boolean nextPermutation(List<Integer> arr) {
        int n = arr.size();
        int i = n - 2;
        
        while (i >= 0 && arr.get(i) >= arr.get(i + 1)) {
            i--;
        }
        
        if (i < 0) return false;
        
        int j = n - 1;
        while (arr.get(j) <= arr.get(i)) {
            j--;
        }
        
        Collections.swap(arr, i, j);
        Collections.reverse(arr.subList(i + 1, n));
        
        return true;
    }
    
    static class TSPResult {
        List<Integer> path;
        int cost;
        int nodesExplored;
        int nodesPruned;
        double executionTime;
        List<String> steps;
        
        TSPResult(List<Integer> path, int cost, int nodes, int pruned, 
                 double time, List<String> steps) {
            this.path = path;
            this.cost = cost;
            this.nodesExplored = nodes;
            this.nodesPruned = pruned;
            this.executionTime = time;
            this.steps = steps;
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Traveling Salesman Problem - Branch & Bound ===\n");
        
        // Test case 1: Small symmetric graph
        int[][] graph1 = {
            {0, 10, 15, 20},
            {10, 0, 35, 25},
            {15, 35, 0, 30},
            {20, 25, 30, 0}
        };
        
        System.out.println("Test Case 1: 4-city symmetric TSP");
        System.out.println("Graph:");
        printGraph(graph1);
        
        TravelingSalesman tsp1 = new TravelingSalesman(graph1, true);
        TSPResult result1 = tsp1.solveTSP();
        
        System.out.println("\nBranch & Bound Solution:");
        for (String step : result1.steps) {
            System.out.println(step);
        }
        
        // Compare with brute force
        System.out.println("\nComparing with Brute Force:");
        TSPResult bruteResult1 = tsp1.bruteForceTSP();
        for (String step : bruteResult1.steps) {
            System.out.println(step);
        }
        
        System.out.println("\nEfficiency Comparison:");
        System.out.println("Branch & Bound: " + result1.nodesExplored + " nodes explored, " 
                         + result1.nodesPruned + " pruned");
        System.out.println("Brute Force: " + bruteResult1.nodesExplored + " permutations checked");
        System.out.println("Speedup: " + String.format("%.2fx", 
            bruteResult1.executionTime / result1.executionTime));
        
        // Test case 2: Larger asymmetric graph
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Test Case 2: 5-city asymmetric TSP");
        
        int[][] graph2 = {
            {0, 12, 29, 22, 13},
            {12, 0, 19, 28, 25},
            {29, 19, 0, 21, 10},
            {22, 28, 21, 0, 24},
            {13, 25, 10, 24, 0}
        };
        
        System.out.println("Graph:");
        printGraph(graph2);
        
        TravelingSalesman tsp2 = new TravelingSalesman(graph2, false);
        TSPResult result2 = tsp2.solveTSP();
        
        System.out.println("\nBranch & Bound Results:");
        System.out.println("Best path: " + result2.path);
        System.out.println("Best cost: " + result2.cost);
        System.out.println("Nodes explored: " + result2.nodesExplored);
        System.out.println("Nodes pruned: " + result2.nodesPruned);
        System.out.println("Execution time: " + String.format("%.2f", result2.executionTime) + " ms");
        
        // Performance analysis
        System.out.println("\n=== Performance Analysis ===");
        System.out.println("Branch & Bound Advantages:");
        System.out.println("- Optimal solution guaranteed");
        System.out.println("- Systematic pruning reduces search space");
        System.out.println("- Best-first search finds good solutions early");
        System.out.println("- Memory efficient compared to dynamic programming");
        
        System.out.println("\nWhen to use Branch & Bound for TSP:");
        System.out.println("- Small to medium problem sizes (< 20 cities)");
        System.out.println("- When optimal solution is required");
        System.out.println("- When good bounds can be computed efficiently");
        System.out.println("- Problems with strong constraint structure");
        
        demonstrateScaling();
    }
    
    private static void printGraph(int[][] graph) {
        int n = graph.length;
        System.out.print("    ");
        for (int i = 0; i < n; i++) {
            System.out.printf("%4d", i);
        }
        System.out.println();
        
        for (int i = 0; i < n; i++) {
            System.out.printf("%2d: ", i);
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 0) {
                    System.out.print("   -");
                } else {
                    System.out.printf("%4d", graph[i][j]);
                }
            }
            System.out.println();
        }
    }
    
    private static void demonstrateScaling() {
        System.out.println("\n=== Scaling Analysis ===");
        
        Random rand = new Random(42);
        int[] sizes = {4, 5, 6, 7};
        
        System.out.println("Problem Size | Nodes Explored | Nodes Pruned | Time (ms) | Pruning %");
        System.out.println("-".repeat(70));
        
        for (int size : sizes) {
            // Generate random symmetric graph
            int[][] graph = new int[size][size];
            for (int i = 0; i < size; i++) {
                for (int j = i + 1; j < size; j++) {
                    int dist = 10 + rand.nextInt(40);
                    graph[i][j] = graph[j][i] = dist;
                }
            }
            
            TravelingSalesman tsp = new TravelingSalesman(graph, false);
            TSPResult result = tsp.solveTSP();
            
            double pruningPercent = (double) result.nodesPruned / 
                                  (result.nodesExplored + result.nodesPruned) * 100;
            
            System.out.printf("%11d | %13d | %11d | %8.2f | %8.1f%%\n",
                size, result.nodesExplored, result.nodesPruned, 
                result.executionTime, pruningPercent);
        }
        
        System.out.println("\nKey Observations:");
        System.out.println("- Pruning effectiveness increases with problem size");
        System.out.println("- Branch & bound dramatically reduces search space");
        System.out.println("- Good bounds are crucial for performance");
    }
}